File: crates/r4s_engine/src/roll_flow.rs
//! Máquina de estados de una tirada:
//!
//! ```text
//! declarada ──(DM)──► aprobada ──(DM)──► oposicion ──(jugador)──► tirada ──► resuelta
//!     ├──(DM)──► contraoferta ──(jugador acepta / edita)──► declarada
//!     ├──(DM)──► rechazada ────(jugador edita)────────────► declarada
//!     ├──(DM)──► sin_tirada ──► resuelta (narración directa)
//!     └──(jugador)──► retirada (también desde contraoferta o rechazada)
//! ```

use crate::dice::DiceRoll;
use crate::error::EngineError;
use crate::resolve::{Outcome, resolve};
use crate::types::{Character, Skill, TieWinner};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum RollState {
    Declarada,
    Aprobada,
    Contraoferta,
    Rechazada,
    SinTirada,
    Oposicion,
    Tirada,
    Resuelta,
    Retirada,
}

impl RollState {
    pub const ALL: [RollState; 9] = [
        RollState::Declarada,
        RollState::Aprobada,
        RollState::Contraoferta,
        RollState::Rechazada,
        RollState::SinTirada,
        RollState::Oposicion,
        RollState::Tirada,
        RollState::Resuelta,
        RollState::Retirada,
    ];

    /// Identificador que se guarda en Firestore.
    pub fn as_str(self) -> &'static str {
        match self {
            RollState::Declarada => "declarada",
            RollState::Aprobada => "aprobada",
            RollState::Contraoferta => "contraoferta",
            RollState::Rechazada => "rechazada",
            RollState::SinTirada => "sin_tirada",
            RollState::Oposicion => "oposicion",
            RollState::Tirada => "tirada",
            RollState::Resuelta => "resuelta",
            RollState::Retirada => "retirada",
        }
    }

    pub fn parse(s: &str) -> Option<Self> {
        Self::ALL.into_iter().find(|st| st.as_str() == s)
    }

    pub fn is_terminal(self) -> bool {
        matches!(self, RollState::Resuelta | RollState::Retirada)
    }
}

/// Quién intenta la transición, relativo a la tirada.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Actor {
    Dm,
    /// Dueño del personaje que declaró la acción.
    Owner,
    /// Cualquier otro miembro de la sala.
    Other,
}

/// Referencia a una habilidad del personaje en el momento de declarar.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SkillRef {
    pub index: usize,
    pub name: String,
    pub level: u8,
}

impl SkillRef {
    pub fn from_character(character: &Character, index: usize) -> Result<Self, EngineError> {
        let skill: &Skill = character
            .skills
            .get(index)
            .ok_or(EngineError::SkillIndexOutOfRange(index))?;
        Ok(Self {
            index,
            name: skill.name.clone(),
            level: skill.level,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RollResult {
    Exito,
    Fallo,
    Narrado,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AdvanceState {
    Pendiente,
    Aplicado,
    NoAplica,
}

/// Lo que se aplicó a la hoja al cerrar la tirada.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct AppliedAdvance {
    pub xp_gained: u32,
    pub xp_spent: u32,
    pub new_skill: Option<Skill>,
    pub replaced_index: Option<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HistoryEntry {
    pub from: Option<RollState>,
    pub to: RollState,
    pub by: Actor,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RollRecord {
    pub state: RollState,
    pub action: String,
    pub skill: SkillRef,
    pub counter_offer: Option<SkillRef>,
    pub dm_note: Option<String>,
    pub opposition: Option<DiceRoll>,
    pub player_roll: Option<DiceRoll>,
    pub result: Option<RollResult>,
    pub narration: Option<String>,
    pub tie_winner: Option<TieWinner>,
    pub advance: Option<AdvanceState>,
    pub applied: Option<AppliedAdvance>,
    pub history: Vec<HistoryEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum FlowAction {
    Approve,
    CounterOffer { skill: SkillRef, note: Option<String> },
    Reject { note: Option<String> },
    Narrate { narration: String },
    AcceptCounterOffer,
    Redeclare { action: String, skill: SkillRef },
    Withdraw,
    RollOpposition { dice: DiceRoll },
    RollPlayer { dice: DiceRoll },
    Resolve { tie_winner: TieWinner },
    ApplyAdvance(AppliedAdvance),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FlowActionKind {
    Approve,
    CounterOffer,
    Reject,
    Narrate,
    AcceptCounterOffer,
    Redeclare,
    Withdraw,
    RollOpposition,
    RollPlayer,
    Resolve,
    ApplyAdvance,
}

impl FlowActionKind {
    pub const ALL: [FlowActionKind; 11] = [
        FlowActionKind::Approve,
        FlowActionKind::CounterOffer,
        FlowActionKind::Reject,
        FlowActionKind::Narrate,
        FlowActionKind::AcceptCounterOffer,
        FlowActionKind::Redeclare,
        FlowActionKind::Withdraw,
        FlowActionKind::RollOpposition,
        FlowActionKind::RollPlayer,
        FlowActionKind::Resolve,
        FlowActionKind::ApplyAdvance,
    ];
}

impl FlowAction {
    pub fn kind(&self) -> FlowActionKind {
        match self {
            FlowAction::Approve => FlowActionKind::Approve,
            FlowAction::CounterOffer { .. } => FlowActionKind::CounterOffer,
            FlowAction::Reject { .. } => FlowActionKind::Reject,
            FlowAction::Narrate { .. } => FlowActionKind::Narrate,
            FlowAction::AcceptCounterOffer => FlowActionKind::AcceptCounterOffer,
            FlowAction::Redeclare { .. } => FlowActionKind::Redeclare,
            FlowAction::Withdraw => FlowActionKind::Withdraw,
            FlowAction::RollOpposition { .. } => FlowActionKind::RollOpposition,
            FlowAction::RollPlayer { .. } => FlowActionKind::RollPlayer,
            FlowAction::Resolve { .. } => FlowActionKind::Resolve,
            FlowAction::ApplyAdvance(_) => FlowActionKind::ApplyAdvance,
        }
    }
}

fn clean_text(s: &str) -> String {
    s.trim().to_string()
}

fn clean_note(note: Option<String>) -> Option<String> {
    note.map(|n| clean_text(&n)).filter(|n| !n.is_empty())
}

/// Tabla de permisos: estado actual × acción × actor.
pub fn permits(state: RollState, kind: FlowActionKind, actor: Actor) -> bool {
    use Actor::*;
    use FlowActionKind as K;
    use RollState as S;
    match (state, kind) {
        (S::Declarada, K::Approve | K::CounterOffer | K::Reject | K::Narrate) => actor == Dm,
        (S::Contraoferta, K::AcceptCounterOffer) => actor == Owner,
        (S::Declarada | S::Contraoferta | S::Rechazada, K::Redeclare | K::Withdraw) => actor == Owner,
        (S::Aprobada, K::RollOpposition) => actor == Dm,
        (S::Oposicion, K::RollPlayer) => actor == Owner,
        (S::Tirada, K::Resolve) => matches!(actor, Owner | Dm),
        (S::Resuelta, K::ApplyAdvance) => matches!(actor, Owner | Dm),
        _ => false,
    }
}

/// Acciones que `actor` puede hacer ahora mismo sobre la tirada.
pub fn allowed_actions(record: &RollRecord, actor: Actor) -> Vec<FlowActionKind> {
    FlowActionKind::ALL
        .into_iter()
        .filter(|&k| permits(record.state, k, actor))
        .filter(|&k| k != FlowActionKind::ApplyAdvance || record.advance == Some(AdvanceState::Pendiente))
        .collect()
}

/// Crea una tirada en estado `declarada`.
pub fn declare(action: &str, skill: SkillRef) -> Result<RollRecord, EngineError> {
    let action = clean_text(action);
    if action.is_empty() {
        return Err(EngineError::EmptyAction);
    }
    Ok(RollRecord {
        state: RollState::Declarada,
        action,
        skill,
        counter_offer: None,
        dm_note: None,
        opposition: None,
        player_roll: None,
        result: None,
        narration: None,
        tie_winner: None,
        advance: None,
        applied: None,
        history: vec![HistoryEntry {
            from: None,
            to: RollState::Declarada,
            by: Actor::Owner,
        }],
    })
}

/// Aplica una acción a la tirada y devuelve la tirada nueva.
pub fn transition(record: &RollRecord, action: FlowAction, actor: Actor) -> Result<RollRecord, EngineError> {
    let kind = action.kind();
    if !permits(record.state, kind, actor) {
        return Err(EngineError::TransitionNotAllowed {
            state: record.state,
            action: kind,
            actor,
        });
    }

    let mut next = record.clone();
    let goto = |next: &mut RollRecord, to: RollState| {
        next.history.push(HistoryEntry {
            from: Some(next.state),
            to,
            by: actor,
        });
        next.state = to;
    };

    match action {
        FlowAction::Approve => goto(&mut next, RollState::Aprobada),
        FlowAction::CounterOffer { skill, note } => {
            next.counter_offer = Some(skill);
            next.dm_note = clean_note(note);
            goto(&mut next, RollState::Contraoferta);
        }
        FlowAction::Reject { note } => {
            next.dm_note = clean_note(note);
            goto(&mut next, RollState::Rechazada);
        }
        FlowAction::Narrate { narration } => {
            let narration = clean_text(&narration);
            if narration.is_empty() {
                return Err(EngineError::EmptyNarration);
            }
            next.narration = Some(narration);
            next.result = Some(RollResult::Narrado);
            next.advance = Some(AdvanceState::NoAplica);
            goto(&mut next, RollState::SinTirada);
            goto(&mut next, RollState::Resuelta);
        }
        FlowAction::AcceptCounterOffer => {
            next.skill = next.counter_offer.take().ok_or(EngineError::TransitionNotAllowed {
                state: record.state,
                action: kind,
                actor,
            })?;
            next.dm_note = None;
            goto(&mut next, RollState::Declarada);
        }
        FlowAction::Redeclare { action, skill } => {
            let action = clean_text(&action);
            if action.is_empty() {
                return Err(EngineError::EmptyAction);
            }
            next.action = action;
            next.skill = skill;
            next.counter_offer = None;
            next.dm_note = None;
            goto(&mut next, RollState::Declarada);
        }
        FlowAction::Withdraw => goto(&mut next, RollState::Retirada),
        FlowAction::RollOpposition { dice } => {
            next.opposition = Some(dice);
            goto(&mut next, RollState::Oposicion);
        }
        FlowAction::RollPlayer { dice } => {
            let expected = next.skill.level as usize;
            if dice.len() != expected {
                return Err(EngineError::DiceCountMismatch {
                    expected,
                    got: dice.len(),
                });
            }
            next.player_roll = Some(dice);
            goto(&mut next, RollState::Tirada);
        }
        FlowAction::Resolve { tie_winner } => {
            let (Some(player), Some(opposition)) = (&next.player_roll, &next.opposition) else {
                unreachable!("tirada y oposición existen en estado Tirada");
            };
            let outcome = resolve(player, opposition, tie_winner);
            next.result = Some(match outcome.outcome {
                Outcome::Success => RollResult::Exito,
                Outcome::Failure => RollResult::Fallo,
            });
            next.tie_winner = Some(tie_winner);
            next.advance = Some(AdvanceState::Pendiente);
            goto(&mut next, RollState::Resuelta);
        }
        FlowAction::ApplyAdvance(applied) => {
            if next.advance != Some(AdvanceState::Pendiente) {
                return Err(EngineError::AdvancementNotPending);
            }
            let expected_xp = u32::from(next.result == Some(RollResult::Fallo));
            if applied.xp_gained != expected_xp {
                return Err(EngineError::NotEligible);
            }
            if applied.new_skill.is_none() && (applied.xp_spent != 0 || applied.replaced_index.is_some()) {
                return Err(EngineError::NotEligible);
            }
            next.applied = Some(applied);
            next.advance = Some(AdvanceState::Aplicado);
        }
    }

    Ok(next)
}

File: web/src/engine/rollFlow.ts
/* Máquina de estados de una tirada:
 *
 *   declarada ──(DM)──► aprobada ──(DM)──► oposicion ──(jugador)──► tirada ──► resuelta
 *       ├──(DM)──► contraoferta ──(jugador acepta / edita)──► declarada
 *       ├──(DM)──► rechazada ────(jugador edita)────────────► declarada
 *       ├──(DM)──► sin_tirada ──► resuelta (narración directa)
 *       └──(jugador)──► retirada (también desde contraoferta o rechazada)
 *
 * Los ids de estado son los que se guardan en Firestore (`estado`, `historial.de/a`). */

import type { DiceRoll } from './dice';
import { EngineError } from './errors';
import { resolve } from './resolve';
import { DIFFICULTIES, MAX_FIXED_TARGET, MAX_MODIFIER, type Character, type Difficulty, type Skill, type TieWinner } from './types';

export const ROLL_STATES = ['declarada', 'aprobada', 'contraoferta', 'rechazada', 'sin_tirada', 'oposicion', 'tirada', 'resuelta', 'retirada'] as const;
export type RollState = (typeof ROLL_STATES)[number];

export function parseRollState(s: unknown): RollState | null {
  return typeof s === 'string' && (ROLL_STATES as readonly string[]).includes(s) ? (s as RollState) : null;
}

export function isTerminal(state: RollState): boolean {
  return state === 'resuelta' || state === 'retirada';
}

/** Quién intenta la transición, relativo a la tirada: el DM, el dueño del personaje o cualquier otro miembro. */
export type Actor = 'dm' | 'owner' | 'other';

/** Referencia a una habilidad del personaje en el momento de declarar. */
export interface SkillRef {
  index: number;
  name: string;
  level: number;
}

export function skillRefFrom(character: Character, index: number): SkillRef {
  const skill: Skill | undefined = character.skills[index];
  if (!skill) throw new EngineError({ kind: 'SkillIndexOutOfRange', index });
  return { index, name: skill.name, level: skill.level };
}

/** La oposición del DM: dados tirados o un objetivo fijo (dificultad estática: 3, 6, 9, 12…). */
export type Opposition = { kind: 'dice'; dice: DiceRoll } | { kind: 'fixed'; target: number };

export function oppositionTotal(o: Opposition): number {
  return o.kind === 'dice' ? o.dice.total() : o.target;
}

/** Objetivo fijo validado: entero entre 1 y `MAX_FIXED_TARGET` (las reglas acotan además por `maxDice`). */
export function fixedOpposition(target: number): Opposition {
  if (!Number.isInteger(target) || target < 1 || target > MAX_FIXED_TARGET) {
    throw new EngineError({ kind: 'InvalidTarget', max: MAX_FIXED_TARGET });
  }
  return { kind: 'fixed', target };
}

/** La dificultad de la tabla a la que corresponde una oposición (por número de dados o por objetivo), si alguna. */
export function difficultyOf(o: Opposition): Difficulty | null {
  return DIFFICULTIES.find((d) => (o.kind === 'dice' ? d.dice === o.dice.length : d.target === o.target)) ?? null;
}

export type RollResult = 'exito' | 'fallo' | 'empate' | 'narrado';
export type AdvanceState = 'pendiente' | 'aplicado' | 'no_aplica';

/** Lo que se aplicó a la hoja al cerrar la tirada. */
export interface AppliedAdvance {
  xpGained: number;
  xpSpent: number;
  newSkill: Skill | null;
  replacedIndex: number | null;
}

export interface HistoryEntry {
  from: RollState | null;
  to: RollState;
  by: Actor;
}

export interface RollRecord {
  state: RollState;
  action: string;
  /** El "para qué" de la acción (opcional): completa la frase «intenta [acción] para [propósito]». */
  purpose: string | null;
  skill: SkillRef;
  counterOffer: SkillRef | null;
  dmNote: string | null;
  opposition: Opposition | null;
  /** Suma de los estados que el DM aplicó al oponer (0 si ninguno) y su lista, para mostrarla. */
  modifier: number;
  modifierNote: string | null;
  playerRoll: DiceRoll | null;
  result: RollResult | null;
  narration: string | null;
  tieWinner: TieWinner | null;
  advance: AdvanceState | null;
  applied: AppliedAdvance | null;
  history: HistoryEntry[];
}

export type FlowAction =
  | { kind: 'approve' }
  | { kind: 'counterOffer'; skill: SkillRef; note: string | null }
  | { kind: 'reject'; note: string | null }
  | { kind: 'narrate'; narration: string }
  | { kind: 'acceptCounterOffer' }
  | { kind: 'redeclare'; action: string; purpose: string | null; skill: SkillRef }
  | { kind: 'withdraw' }
  | { kind: 'rollOpposition'; opposition: Opposition; modifier?: number; modifierNote?: string | null }
  | { kind: 'rollPlayer'; dice: DiceRoll }
  | { kind: 'resolve'; tieWinner: TieWinner }
  | { kind: 'applyAdvance'; applied: AppliedAdvance };

export type FlowActionKind = FlowAction['kind'];

export const FLOW_ACTION_KINDS: readonly FlowActionKind[] = [
  'approve',
  'counterOffer',
  'reject',
  'narrate',
  'acceptCounterOffer',
  'redeclare',
  'withdraw',
  'rollOpposition',
  'rollPlayer',
  'resolve',
  'applyAdvance',
];

function cleanNote(note: string | null): string | null {
  const n = note?.trim() ?? '';
  return n === '' ? null : n;
}

/** Tabla de permisos: estado actual × acción × actor. */
export function permits(state: RollState, kind: FlowActionKind, actor: Actor): boolean {
  switch (kind) {
    case 'approve':
    case 'counterOffer':
    case 'reject':
    case 'narrate':
      return state === 'declarada' && actor === 'dm';
    case 'acceptCounterOffer':
      return state === 'contraoferta' && actor === 'owner';
    case 'redeclare':
    case 'withdraw':
      return (state === 'declarada' || state === 'contraoferta' || state === 'rechazada') && actor === 'owner';
    case 'rollOpposition':
      return state === 'aprobada' && actor === 'dm';
    case 'rollPlayer':
      return state === 'oposicion' && actor === 'owner';
    case 'resolve':
      return state === 'tirada' && (actor === 'owner' || actor === 'dm');
    case 'applyAdvance':
      return state === 'resuelta' && (actor === 'owner' || actor === 'dm');
  }
}

/** Acciones que `actor` puede hacer ahora mismo sobre una tirada en `state` con avance `advance`. */
export function allowedActionsFor(state: RollState, advance: AdvanceState | null, actor: Actor): FlowActionKind[] {
  return FLOW_ACTION_KINDS.filter((k) => permits(state, k, actor)).filter((k) => k !== 'applyAdvance' || advance === 'pendiente');
}

export function allowedActions(record: Pick<RollRecord, 'state' | 'advance'>, actor: Actor): FlowActionKind[] {
  return allowedActionsFor(record.state, record.advance, actor);
}

/** Crea una tirada en estado `declarada`. */
export function declare(action: string, skill: SkillRef, purpose: string | null = null): RollRecord {
  const text = action.trim();
  if (text === '') throw new EngineError({ kind: 'EmptyAction' });
  return {
    state: 'declarada',
    action: text,
    purpose: cleanNote(purpose),
    skill,
    counterOffer: null,
    dmNote: null,
    opposition: null,
    modifier: 0,
    modifierNote: null,
    playerRoll: null,
    result: null,
    narration: null,
    tieWinner: null,
    advance: null,
    applied: null,
    history: [{ from: null, to: 'declarada', by: 'owner' }],
  };
}

/** Aplica una acción a la tirada y devuelve la tirada nueva (la original no cambia). */
export function transition(record: RollRecord, action: FlowAction, actor: Actor): RollRecord {
  const kind = action.kind;
  const notAllowed = () => new EngineError({ kind: 'TransitionNotAllowed', state: record.state, action: kind, actor });
  if (!permits(record.state, kind, actor)) throw notAllowed();

  const next: RollRecord = { ...record, history: [...record.history] };
  const goto = (to: RollState) => {
    next.history.push({ from: next.state, to, by: actor });
    next.state = to;
  };

  switch (action.kind) {
    case 'approve':
      goto('aprobada');
      break;
    case 'counterOffer':
      next.counterOffer = action.skill;
      next.dmNote = cleanNote(action.note);
      goto('contraoferta');
      break;
    case 'reject':
      next.dmNote = cleanNote(action.note);
      goto('rechazada');
      break;
    case 'narrate': {
      const narration = action.narration.trim();
      if (narration === '') throw new EngineError({ kind: 'EmptyNarration' });
      next.narration = narration;
      next.result = 'narrado';
      next.advance = 'no_aplica';
      goto('sin_tirada');
      goto('resuelta');
      break;
    }
    case 'acceptCounterOffer': {
      if (!next.counterOffer) throw notAllowed();
      next.skill = next.counterOffer;
      next.counterOffer = null;
      next.dmNote = null;
      goto('declarada');
      break;
    }
    case 'redeclare': {
      const text = action.action.trim();
      if (text === '') throw new EngineError({ kind: 'EmptyAction' });
      next.action = text;
      next.purpose = cleanNote(action.purpose);
      next.skill = action.skill;
      next.counterOffer = null;
      next.dmNote = null;
      goto('declarada');
      break;
    }
    case 'withdraw':
      goto('retirada');
      break;
    case 'rollOpposition': {
      const modifier = action.modifier ?? 0;
      if (!Number.isInteger(modifier) || Math.abs(modifier) > MAX_MODIFIER) throw new EngineError({ kind: 'InvalidModifier', max: MAX_MODIFIER });
      next.opposition = action.opposition.kind === 'fixed' ? fixedOpposition(action.opposition.target) : action.opposition;
      next.modifier = modifier;
      next.modifierNote = modifier === 0 ? null : cleanNote(action.modifierNote ?? null);
      goto('oposicion');
      break;
    }
    case 'rollPlayer': {
      const expected = next.skill.level;
      if (action.dice.length !== expected) throw new EngineError({ kind: 'DiceCountMismatch', expected, got: action.dice.length });
      next.playerRoll = action.dice;
      goto('tirada');
      break;
    }
    case 'resolve': {
      // En estado `tirada` siempre existen la tirada y la oposición.
      if (!next.playerRoll || !next.opposition) throw notAllowed();
      const outcome = resolve(next.playerRoll, oppositionTotal(next.opposition), action.tieWinner, next.modifier);
      next.result = outcome.outcome === 'success' ? 'exito' : outcome.outcome === 'tie' ? 'empate' : 'fallo';
      next.tieWinner = action.tieWinner;
      next.advance = 'pendiente';
      goto('resuelta');
      break;
    }
    case 'applyAdvance': {
      if (next.advance !== 'pendiente') throw new EngineError({ kind: 'AdvancementNotPending' });
      const { applied } = action;
      const expectedXp = next.result === 'fallo' ? 1 : 0;
      if (applied.xpGained !== expectedXp) throw new EngineError({ kind: 'NotEligible' });
      if (applied.newSkill === null && (applied.xpSpent !== 0 || applied.replacedIndex !== null)) {
        throw new EngineError({ kind: 'NotEligible' });
      }
      next.applied = applied;
      next.advance = 'aplicado';
      break;
    }
  }

  return next;
}
